File: src/Apps/NetPad.Web/BackgroundServices/QueryBackgroundService.cs
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Threading;
using System.Threading.Tasks;
using ElectronNET.API;
using NetPad.Common;
using NetPad.Queries;
using NetPad.Sessions;

namespace NetPad.BackgroundServices
{
    public class QueryBackgroundService : BackgroundService
    {
        private readonly ISession _session;
        private readonly Settings _settings;

        public QueryBackgroundService(ISession session, Settings settings)
        {
            _session = session;
            _settings = settings;
        }

        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            ReactToQueryPropertyChanged();

            // Electron.IpcMain.RemoveAllListeners("save-query");
            // Electron.IpcMain.On("save-query", async (msg) =>
            // {
            //
            // });
        }

        private void ReactToQueryPropertyChanged()
        {
            _session.OpenQueries.CollectionChanged += (_,  changes) =>
            {
                if (changes.Action == NotifyCollectionChangedAction.Add && changes.NewItems?.Count > 0)
                {
                    foreach (Query query in changes.NewItems)
                    {
                        query.OnPropertyChanged.Add(async (args) =>
                        {
                            Electron.IpcMain.Send(BrowserWindow, "query-property-changed", Serialize(new
                            {
                                QueryId = query.Id,
                                PropertyName = args.PropertyName,
                                NewValue = args.NewValue
                            }));
                        });
                    }
                }
                else if (changes.Action == NotifyCollectionChangedAction.Remove)
                {
                    var queries = changes.OldItems as IList<Query>;
                    if (queries == null) return;

                    foreach (var query in queries)
                    {
                        query.RemoveAllPropertyChangedHandlers();
                    }
                }
            };
        }
    }
}

File: src/Apps/NetPad.Web/App/src/main-window/background-services/query-background-service.ts
import {ISession} from "@domain";
import {ipcRenderer} from "electron";
import {IBackgroundService} from "./ibackground-service";

export class QueryBackgroundService implements IBackgroundService{
    constructor(@ISession readonly session: ISession) {}

    public start(): Promise<void> {
        ipcRenderer.on("query-property-changed", (event, json) =>
        {
            const update = JSON.parse(json);
            const query = this.session.queries.find(s => s.id == update.queryId);
            const propName = update.propertyName.charAt(0).toLowerCase() + update.propertyName.slice(1);
            query[propName] = update.newValue;
        });

        return Promise.resolve();
    }

    public stop(): void {
        ipcRenderer.removeAllListeners("query-property-changed");
    }
}
